Feature/formatters

NikolayKorsakov opened the pull request:
Добавлен клиент, поддерживающий форматтеры.

NikolayKorsakov committed:
added formatter client
added first test for it

Ewsf0r committed:
Добавлен тестовый сервис

NikolayKorsakov committed:
Поправлен проект с примером работы клиента с форматтером
Поправлены тесты
Добавлены тесты для protobuf, message pack и json форматтеров

Ewsf0r committed:
fixed formatter client for protobuf

Ewsf0r committed:
readme updated

NikolayKorsakov committed:
Тесты от разных форматтеров перенесены в одни файл
В клиенте унифицирован код для запросов с получением результатов и без оного
Добавлен стресс тест для записи/чтения 10 тысяч элементов

NikolayKorsakov committed:
Добавлено тестирование методов пут и патч
В клиента с форматером добавлены метод пут без возвращаемого значения и метод патч с возвращаемым значением

NikolayKorsakov committed:
в базового клиента добавлены пуш и патч запросы
в базовом клиенте гет запросы переделаны через call
 в форматированном клиенте к методам, скрывающим таковые у базового добавлено ключевое слово new
 в тестах переименована переменная - адрес сервера

litichevskiydv commented on .github/workflows/push.yml:
```diff
@@ -22,8 +22,8 @@ jobs:
           dotnet-version: | 
-            2.1.x
-            6.0.x
+            2.0.x
```
Для чего нам нужен .NET Core второй версии?

litichevskiydv commented on .github/workflows/release.yml:
```diff
@@ -17,8 +17,8 @@ jobs:
           dotnet-version: | 
-            2.1.x
-            6.0.x
+            2.0.x
```
Для чего нам нужен .NET Core второй версии?

litichevskiydv commented on README.md:
```diff
@@ -14,16 +14,27 @@ dotnet add package Byndyusoft.ApiClient
 
 ## Usage
-To create an API client:
+To create an Base API client:
```
Мы создаем клиент конкретного API `SomeApiClient`, поэтому слово `Base` здесь лишнее

litichevskiydv commented on README.md:
````diff
@@ -14,16 +14,27 @@ dotnet add package Byndyusoft.ApiClient
+```
+public class SomeApiClient : FormatterClient
+{
+	public SomeApiClient(HttpClient client, MediaTypeFormatter formatter, IOptions<ApiClientSettings> apiSettings) : base(client, formatter, apiSettings)
````
Экземпляр клиента `SomeApiClient` потребитель получает из IoC, следовательно и реализация `MediaTypeFormatter` должна быть зарегана в IoC?
Что будет, если я захочу в приложении использовать форматеры и для Protobuf и для MessagePack, как будет работать резолв зависимостей?

litichevskiydv replied:
Кроме того, клиент должен реализовывать контракт API, к которому он обращается

NikolayKorsakov replied:
+

litichevskiydv commented on README.md:
````diff
@@ -14,16 +14,27 @@ dotnet add package Byndyusoft.ApiClient
 ```
+or create your API client class and derive it from FormatterClient class:
+```
+public class SomeApiClient : FormatterClient
````
Зачем нам ещё один базовый класс?

litichevskiydv commented on example/Client/TestFormatterClient.cs:
```diff
@@ -0,0 +1,52 @@
+public class TestFormatterClient : FormatterClient
+{
+    public TestFormatterClient
+        (
```
Не переносить скобку на следующую строку

litichevskiydv commented on example/Client/TestFormatterClient.cs:
```diff
@@ -0,0 +1,52 @@
+            IOptions<ApiClientSettings> apiSettings
+        )
+        : base
+        (
```
Вытянуть вырражение в одну строку и отформатировать

litichevskiydv commented on example/Client/TestFormatterClient.cs:
```diff
@@ -0,0 +1,52 @@
+    }
+
+
+    public Task<TResult> GetAsync<TResult>(string url, CancellationToken cancellationToken)
```
Смысл клиента API в том, что в нем уже прописаны роуты и они принимают конкретные типы данных, посмотри пример в шаблоне.
Здесь же и роут приходит параметром и тип возвращаемых данных произвольный, это нужно переделать.

litichevskiydv commented on example/Client/TestFormatterClient.cs:
```diff
@@ -0,0 +1,52 @@
+    public Task<TResult> GetAsync<TResult>(string url, CancellationToken cancellationToken)
+        => base.GetAsync<TResult>(url, cancellationToken);
+
+    public Task<TResult> GetAsync<TParams, TResult>(string url, CancellationToken cancellationToken, TParams? dto = null)
```
Смысл клиента API в том, что в нем уже прописаны роуты и они принимают конкретные типы данных, посмотри пример в шаблоне.
Здесь же и роут приходит параметром и типы передаваемых и возвращаемых данных произвольны, это нужно переделать. Аналогичная история со всеми остальными методами клиента

litichevskiydv commented on example/Controllers/TestController.cs:
```diff
@@ -0,0 +1,134 @@
+                new ApiClientSettings(){ConnectionString = connectionString}
+            );
+
+            var formatterClient = new TestFormatterClient(httpClient, formatter, options);
```
Клиенты должны извлекаться из IoC для удобства их использования потребителем. Аналогично для остальных методов.

litichevskiydv commented on example/Controllers/TestController.cs:
```diff
@@ -0,0 +1,134 @@
+                    10101,
+                    "tuple",
+                    null));
+            result = await formatterClient.PostAsync<SimpleModel>("/post", result, cancellationToken);
```
Получается, что приложение само себя вызывает?

litichevskiydv commented on example/Controllers/TestController.cs:
```diff
@@ -0,0 +1,134 @@
+    {
+        [HttpGet("test_protobuf")]
+        [FormatFilter]
+        public async Task<IActionResult> TestProtoBuf(CancellationToken cancellationToken)
```
Это именно тестирование того, что форматеры работают? Если так, то нужно написать именно тест, в котором вызывается метод приложения. Аналогично для остальных методов.

litichevskiydv commented on example/Models/ParamsTestModel.cs:
```diff
@@ -0,0 +1,6 @@
+public record ParamsTestModel(
+    int property,
+    string field,
+    int? nullable);
```
Перенести `);` на новую строчку и отформатировать

litichevskiydv commented on example/Models/SimpleModel.cs:
```diff
@@ -0,0 +1,42 @@
+[ProtoContract]
+public class SimpleModel
+{
+    [ProtoMember(2)][JsonInclude]  public string Field = default!;
```
Выровняй нумерацию, чтобы она шла по порядку

litichevskiydv commented on example/Swagger/ApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+        /// <summary>
+        ///     UseSwagger
+        /// </summary>
+        public static IApplicationBuilder UseSwaggerWithApiVersionDescriptionProvider(
```
Либо записать в одну строку, либо перенести `)` на новую строчку и отформатировать

litichevskiydv commented on example/Swagger/ApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+                   .UseSwaggerUI(options =>
+                   {
+                       foreach (var apiVersionDescription in apiVersionDescriptionProvider.ApiVersionDescriptions)
+                           options.SwaggerEndpoint($"/swagger/{apiVersionDescription.GroupName}/swagger.json", apiVersionDescription.GroupName.ToUpperInvariant());
```
Перенести `$"/swagger/{apiVersionDescription.GroupName}/swagger.json",` на новую строчку

litichevskiydv commented on example/Swagger/ApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+                   .UseSwaggerUI(options =>
+                   {
+                       foreach (var apiVersionDescription in apiVersionDescriptionProvider.ApiVersionDescriptions)
+                           options.SwaggerEndpoint($"/swagger/{apiVersionDescription.GroupName}/swagger.json", apiVersionDescription.GroupName.ToUpperInvariant());
```
Перенести `);` на новую строчку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on example/Swagger/ApplicationBuilderExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+                       options.DisplayRequestDuration();
+                       options.DefaultModelRendering(ModelRendering.Model);
+                       options.DefaultModelExpandDepth(3);
+                   });
```
Перенести `);` на следующую строчку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on example/Swagger/ConfigureSwaggerOptions.cs:
```diff
@@ -0,0 +1,46 @@
+                        Description = apiVersionDescription.IsDeprecated ? "DEPRECATED" : "",
+                        Title = Assembly.GetExecutingAssembly().GetName().Name
+                    });
+
```
Убрать лишний отступ

NikolayKorsakov replied:
+

litichevskiydv commented on example/Swagger/ConfigureSwaggerOptions.cs:
```diff
@@ -0,0 +1,46 @@
+                        Version = apiVersionDescription.ApiVersion.ToString(),
+                        Description = apiVersionDescription.IsDeprecated ? "DEPRECATED" : "",
+                        Title = Assembly.GetExecutingAssembly().GetName().Name
+                    });
```
Перенести `);` на следующую строчку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on example/Swagger/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+        public static IServiceCollection AddSwagger(this IServiceCollection services)
+        {
+            return
+                services.AddTransient<IConfigureOptions<SwaggerGenOptions>, ConfigureSwaggerOptions>()
```
Заменить на стрелочную функцию для экономии места

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -66,48 +61,30 @@ protected string GetAbsoluteUrl(string url)
 
         protected async Task CallAsync(HttpMethod method, string url, object? content, CancellationToken cancellationToken)
+        {
+            await CallAsyncBase(method, url, content, cancellationToken);
```
Заменить на стрелочную функцию и убррать `async ... await`

litichevskiydv commented on src/Byndyusoft.ApiClient/Byndyusoft.ApiClient.csproj:
```diff
@@ -12,13 +12,14 @@
-		<PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
+		<PackageReference Include="Byndyusoft.Net.Http" Version="0.4.4" />
+		<PackageReference Include="Microsoft.Extensions.Options" Version="8.0.2" />
+		<PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
```
Для чего мы используем этот пакет?

litichevskiydv commented on src/Byndyusoft.ApiClient/FormatterClient.cs:
```diff
@@ -0,0 +1,129 @@
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Options;
+
+    public class FormatterClient: BaseClient
```
В Readme задавал вопросы о необходимости отдельного класса, подумай над ними. Как будто его не должно быть.

litichevskiydv commented on src/Byndyusoft.ApiClient/FormatterClient.cs:
```diff
@@ -0,0 +1,129 @@
+            object? content,
+            CancellationToken cancellationToken)
+        {
+            var isProtobuf = _formatter.GetType().Name.ToLower().Contains("protobuf");
```
Так точно не должно быть, контракта должно быть достаточно для работы кода. Посмотри мои эксперименты [здесь](https://github.com/litichevskiydv/WebInfrastructure/blob/develop/src/Infrastructure/Web.Integration/BaseApiClient/BaseClient.cs). Не стоит их в лоб копировать, но в их рамках удалось построить клиент без подобных проверок

litichevskiydv replied:
Разобраться, зачем нужен ` Version = new Version(2, 0)`, возможно, баг [здесь](https://github.com/Byndyusoft/Byndyusoft.AspNetCore.Mvc.Formatters.ProtoBuf/blob/master/src/Formatters/ProtoBufOutputFormatter.cs#L56)

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -66,48 +61,30 @@ protected string GetAbsoluteUrl(string url)
-                    statusCode: response.StatusCode,
-                    content: stringContent);
-            }
+            await Toolkit.EnsureSuccessStatusCode(response);
```
Сделай расширением для `HttpResponseMessage`

litichevskiydv commented on src/Byndyusoft.ApiClient/FormatterClient.cs:
```diff
@@ -0,0 +1,129 @@
+                requestMessage.Version = new Version(1, 0);
+
+            var response = await Client.SendAsync(requestMessage, cancellationToken).ConfigureAwait(false);
+            await Toolkit.EnsureSuccessStatusCode(response).ConfigureAwait(false);
```
Переделать в расширение для `HttpResponseMessage`

litichevskiydv commented on tests/Functional/Client/TestFormatterClient.cs:
```diff
@@ -0,0 +1,58 @@
+using Byndyusoft.ApiClient;
+using Microsoft.Extensions.Options;
+
+public class TestFormatterClient : FormatterClient
```
Клиент должен предоставлять API, для которого он написан. Предлагаю перенести клиент в проект с примерами и там же написать работающий через него тест.

litichevskiydv commented on tests/Functional/Client/TestFormatterClient.cs:
```diff
@@ -0,0 +1,58 @@
+    }
+
+
+    public Task<TResult> GetAsync<TResult>(string url, CancellationToken cancellationToken)
```
Кроме того, методы клиента должны точно соответствовать методам API, для которого он написан, здесь просто копирование методов базового класса

litichevskiydv commented on tests/Functional/Controllers/FormatterController.cs:
```diff
@@ -0,0 +1,46 @@
+
+    [Controller]
+    [Route("formatter")]
+    public class FormatterController : ControllerBase
```
Придумай название, которое коррелирует с названием модели, с которой контролер работает.
Аналогичным образом скорректируй роут.

litichevskiydv commented on tests/Functional/Controllers/FormatterController.cs:
```diff
@@ -0,0 +1,46 @@
+        public IActionResult WithParams(
+            [FromQuery] int property,
+            [FromQuery] string field,
+            [FromQuery] int? nullable)
```
Перенести `)` на следующую строку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on tests/Functional/Controllers/FormatterController.cs:
```diff
@@ -0,0 +1,46 @@
+        [FormatFilter]
+        public IActionResult Post([FromBody] SimpleModel model)
+        {
+            return Ok(model);
```
Заменить на стрелочную функцию

litichevskiydv commented on tests/Functional/Controllers/FormatterController.cs:
```diff
@@ -0,0 +1,46 @@
+        [FormatFilter]
+        public IActionResult Put([FromBody] SimpleModel model)
+        {
+            return Ok(model);
```
Заменить на стрелочную функцию

litichevskiydv commented on tests/Functional/Controllers/FormatterController.cs:
```diff
@@ -0,0 +1,46 @@
+        [FormatFilter]
+        public IActionResult Get()
+        {
+            return Ok(SimpleModel.Create());
```
Заменить на стрелочную функцию

litichevskiydv commented on tests/Functional/Controllers/FormatterController.cs:
```diff
@@ -0,0 +1,46 @@
+            [FromQuery] string field,
+            [FromQuery] int? nullable)
+        {
+            return Ok(SimpleModel.Create(property, field:field, nullable:nullable));
```
Заменить на стрелочную функцию

litichevskiydv commented on tests/Functional/Controllers/FormatterListController.cs:
```diff
@@ -0,0 +1,76 @@
+
+    [Controller]
+    [Route("formatter/list")]
+    public class FormatterListController : ControllerBase
```
Почему для тестов нам нужно два контролера?

NikolayKorsakov replied:
+

litichevskiydv commented on tests/Functional/Models/ParamsTestModel.cs:
```diff
@@ -0,0 +1,6 @@
+public record ParamsTestModel(
+    int property,
+    string field,
+    int? nullable);
```
Перенести `);` на следующую строку и отформатировать

litichevskiydv commented on tests/Functional/Models/SimpleModel.cs:
```diff
@@ -0,0 +1,79 @@
+[ProtoContract]
+public class SimpleModel
+{
+    [ProtoMember(2)][JsonInclude] public string Field = default!;
```
Упорядочить поля по номерам

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+    protected async Task GetAsyncWithParamsTest()
+    {
+        // Arrange
+        var input = new ParamsTestModel(
```
Записать в одну строку

NikolayKorsakov replied:
+

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+        var response = await TestSubject.GetAsync<ParamsTestModel,SimpleModel>(
+            "/formatter/with_params",
+            CancellationToken.None,
+            input);
```
Перенести `);` на следующую строку и отформатировать

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+            "/formatter/with_params",
+            CancellationToken.None,
+            input);
+        // Assert
```
Добавить отступ перед строкой

NikolayKorsakov replied:
+

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+            async () => 
+                await TestSubject.DeleteAsync(
+            $"/formatter/list/delete/{id}",
+            CancellationToken.None));
```
Перенести `)` на новую строку и отформатировать

litichevskiydv replied:
Перенести `);` на новую строку и отформатировать

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+        var input = SimpleModel.Create();
+
+        // Act
+        var response = await TestSubject.PostAsync<SimpleModel>("/formatter/post", input, CancellationToken.None);
```
Клиент должен скрывать роуты, смысл его существования в том, чтобы клиенты не писали их при каждом обращении к API.

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+        var id = Interlocked.Increment(ref _idProvider);
+        testOutputHelper.WriteLine($"id: {id}");
+        ListModel.Data.Add(id, new List<SimpleModel>());
+        // Assert
```
Добавить отступ перед строкой

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+        var input = SimpleModel.Create();
+
+        // Act
+        var response = await TestSubject.PostAsync<SimpleModel>
```
Какой сценарий проверяет этот тест?

NikolayKorsakov replied:
+

litichevskiydv commented on tests/Functional/MvcFormattersTests.cs:
```diff
@@ -0,0 +1,190 @@
+    }
+
+    [Fact]
+    protected async Task ListStressTest()
```
Какой сценарий проверяет этот тест?

litichevskiydv commented on tests/Functional/MvcTestFixture.cs:
```diff
@@ -0,0 +1,87 @@
+        URL = $"http://localhost:{FreeTcpPort()}";
+        _host =
+            Host.CreateDefaultBuilder()
+                .ConfigureWebHostDefaults(webBuilder =>
```
Перенести `webBuilder =>` на слудующую строку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on tests/Functional/MvcTestFixture.cs:
```diff
@@ -0,0 +1,87 @@
+                                              webBuilder.UseUrls(URL);
+                                              webBuilder.ConfigureServices(ConfigureServices);
+                                              webBuilder.Configure(Configure);
+                                          })
```
Перенести `)` на следующую строку и отформатировать

Ewsf0r committed:
renoved one of vlients types

Ewsf0r committed:
merge

Ewsf0r committed:
Все контроллеры и данные перенесены из теста в пример
Поправлены гет-запросы у клиента

Ewsf0r committed:
Запросы в клиенте исправлены

Ewsf0r committed:
Правка версии хттп протокола для поддежки протобафа

Ewsf0r committed:
Добавлены запросы на добавление и удаление списка в проект пример

Ewsf0r committed:
Удален пакет Newtonsoft.Json
Добавлены интерфейсы для клиента
Добавлен тест на несколько контроллеров

Ewsf0r committed:
Поправлено получение клиента в тестах

NikolayKorsakov committed:
Правки по требованиям к ПР

NikolayKorsakov committed:
nuget update

NikolayKorsakov committed:
Removed formatter provider
Changed basic model

Ewsf0r committed:
refactoring of test model, controller, client and tests
found and resolved client delition with parameters bug

Ewsf0r committed:
readme update

Ewsf0r committed:
removed some maintanence variables

Ewsf0r committed:
fixed person model for protobuf test

NikolayKorsakov committed:
Проект примера разнесен на более мелкие

NikolayKorsakov committed:
перенос файла

NikolayKorsakov committed:
fixed test

NikolayKorsakov committed:
formatting

Ewsf0r committed:
Разделение тестов: тесты клиента остались на месте, тесты проекта-примера перенесены
Создание тестового сервера завязано на стартап примера
Бенчмарк так же перенесен к примеру

Ewsf0r committed:
Fixed tests
Fixed benchmark

Ewsf0r committed:
linux paths fix

Ewsf0r committed:
Revert "linux paths fix"

This reverts commit 3e8e09fec5c5d3e108af40ddc6c8c98207c7a767.

Ewsf0r committed:
paths fix

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Client/Byndyusoft.ApiClient.Example.Client.csproj:
```diff
@@ -0,0 +1,15 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Давай название папки `Example` заменим на `example`, чтобы было по аналогии с `src` и `tests`

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Client/PersonModelListClient.cs:
```diff
@@ -0,0 +1,45 @@
+using Microsoft.Extensions.Options;
+using Models;
+
+public class PersonModelListClient : BaseClient, IPersonModelListClient
```
Предлагаю переименвать `IPersonModelListClient` -> `IPersonListsApi` и соответствующщим образом переименовать клиент и контролер.
По модели `PersonModel`, что нам дает суффикс `Model` в ее названии?

NikolayKorsakov replied:
List тоже убрал

NikolayKorsakov replied:
+

Ewsf0r committed:
linux paths fix

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Benchmark/Program.cs:
```diff
@@ -0,0 +1,17 @@
+            var config = DefaultConfig.Instance;
+            var summary = BenchmarkRunner.Run<FormattersBenchmark>(config, args);
+
+            // Use this to select benchmarks from the console:
```
Убрать закомментированный код

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Benchmark/BenchmarkDataType.cs:
```diff
@@ -0,0 +1,8 @@
+namespace Byndyusoft.ApiClient.Example.Benchmark;
+
+public enum BenchmarkDataType
```
Убери префикс `Example/Byndyusoft.ApiClient.Example.` из директорий по аналоогии с шаблоном API, чтобы навигация в папках стала проще

NikolayKorsakov replied:
Не уверен, что понял правильно, но сделал

litichevskiydv replied:
Все верно сделал. Теперь аналогичным образом сократи названия проектов примера

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Byndyusoft.ApiClient.Example.Server.csproj:
```diff
@@ -0,0 +1,31 @@
+	<ItemGroup>
+		<PackageReference Include="Asp.Versioning.Mvc" Version="8.1.0" />
+		<PackageReference Include="Asp.Versioning.Mvc.ApiExplorer" Version="8.1.0" />
+        <PackageReference Include="Byndyusoft.AspNetCore.Instrumentation.Tracing" Version="1.0.0" />
```
Выровнять отступы

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Byndyusoft.ApiClient.Example.Server.csproj:
```diff
@@ -0,0 +1,31 @@
+        <PackageReference Include="OpenTelemetry.Exporter.Jaeger" Version="1.5.1" />
+        <PackageReference Include="OpenTelemetry.Exporter.Prometheus" Version="1.3.0-rc.2" />
+        <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.11.0" />
+		<PackageReference Include="protobuf-net.AspNetCore" Version="3.2.46" />
```
Этот пакет точно нужен?

NikolayKorsakov replied:
Да. Этот пакет работает, в отличии от "Byndyusoft.AspNetCore.Mvc.Formatters.ProtoBuf".

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Byndyusoft.ApiClient.Example.Server.csproj:
```diff
@@ -0,0 +1,31 @@
+	</ItemGroup>
+
+	<ItemGroup>
+	  <ProjectReference Include="..\Byndyusoft.ApiClient.Example.Client\Byndyusoft.ApiClient.Example.Client.csproj" />
```
Получается, что сервер ссылается на свой же клиент?

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Controllers/PersonModelListController.cs:
```diff
@@ -0,0 +1,125 @@
+    [FormatFilter]
+    public IActionResult GetEveryPerson(int id)
+    {
+        if (!PersonModelDictionary.Data.TryGetValue(id, out var value))
```
Замени `!` на ` == false`, это повысит читаемость кода

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Controllers/PersonModelListController.cs:
```diff
@@ -0,0 +1,125 @@
+    [FormatFilter]
+    public IActionResult DeleteList(int id)
+    {
+        if (!PersonModelDictionary.Data.ContainsKey(id))
```
Замени `!` на ` == false`, это повысит читаемость кода

litichevskiydv replied:
Исправь аналогичные места ниже по коду

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Controllers/PersonModelListController.cs:
```diff
@@ -0,0 +1,125 @@
+    {
+        if (!PersonModelDictionary.Data.TryGetValue(id, out var value))
+            return NotFound();
+        return Ok(value);
```
Добавь отступ перед строкой, чтобы блоки не сливались

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Controllers/PersonModelListController.cs:
```diff
@@ -0,0 +1,125 @@
+    {
+        if (!PersonModelDictionary.Data.ContainsKey(id))
+            return NotFound();
+        PersonModelDictionary.Data.Remove(id, out _);
```
Добавь отступ перед строкой, чтобы блоки на сливались

litichevskiydv replied:
Исправь аналогичные места ниже по коду

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Program.cs:
```diff
@@ -0,0 +1,32 @@
+                    .WriteToOpenTelemetry(activityEventBuilder: StructuredActivityEventBuilder.Instance)
+            )
+            .ConfigureServices((context, services) => services.AddOpenTelemetry(serviceName, context.Configuration))
+            .ConfigureWebHostDefaults(webBuilder => webBuilder.UseStartup<Startup>());
```
Мы взяли курс на отказ от `Startup`, давай и здесь его уберем

NikolayKorsakov replied:
Сейчас `Startup` используется в тестах, чтобы подключится к серверу из основного приложения. В кафке так же сделано. Не подскажешь, где можно глянуть, как этого добиться, если избавиться от `Startup`?

litichevskiydv replied:
Почитай про `WebApplicationFactory<TEntryPoint>`, мы можем использовать `WebApplicationFactory<Program>`
https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-8.0

NikolayKorsakov replied:
+
Убрал IDisposable в тесте.
After the tests of the IClassFixture implementation are executed, [TestServer](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.testhost.testserver) and [HttpClient](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient) are disposed when xUnit disposes of the [WebApplicationFactory](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1).

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Properties/launchSettings.json:
```diff
@@ -0,0 +1,13 @@
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      },
+      "applicationUrl": "https://localhost:5001;http://localhost:5000"
```
Давай уберем `https://localhost:5001;`, чтобы при запуске студия не навязывала самоподписанный серт

NikolayKorsakov replied:
+

litichevskiydv commented on Example/Byndyusoft.ApiClient.Example.Server/Startup.cs:
```diff
@@ -0,0 +1,74 @@
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services
+                .AddApiVersioning(
```
Давай использовать аналогичный код из шаблона, чтобы не создавать множество вариаций одного и того же куска

NikolayKorsakov replied:
+

litichevskiydv commented on Example/tests/FormattersTests.cs:
```diff
@@ -0,0 +1,203 @@
+namespace Byndyusoft.ApiClient.Example.Tests;
```
Отбей отступами разные смысловые блоки

NikolayKorsakov replied:
?

litichevskiydv replied:
Во фрагменте 
```C#
await TestSubject.AddListAsync(listId, cancel);
await TestSubject.AddPersonAsync(listId, oldPerson, cancel);
var actualPerson = await TestSubject.ReplacePersonAsync(listId, newPerson, cancel);
var personId = new PersonId(listId, actualPerson.Id);
actualPerson = await TestSubject.GetPersonAsync(personId, cancel);
personId = new PersonId(listId, actualPerson.Id);
await TestSubject.DeletePersonAsync(personId, cancel);
await TestSubject.DeleteListAsync(listId, cancel);
```
выполняются разные операции: созздание списка и добавление в него, замена и получение, удаление из списка и удаление самого списка. Отдели отступами эти группы операций друг от друга. Аналогично пройдись по остальным тестам.

NikolayKorsakov replied:
+

litichevskiydv commented on Example/tests/FormattersTests.cs:
```diff
@@ -0,0 +1,203 @@
+
+        // Act
+        await TestSubject.AddListAsync(listId, cancel);
+        var response = await TestSubject.ReplacePersonAsync(listId, newPerson, cancel);
```
По названию переменной `response` не понятно что в ней лежит, переименуй во что-то говорящее

litichevskiydv replied:
Исправь аналогичные места ниже по коду

NikolayKorsakov replied:
+

litichevskiydv commented on Example/tests/FormattersTests.cs:
```diff
@@ -0,0 +1,203 @@
+using Xunit;
+using Xunit.Abstractions;
+
+public abstract class FormattersTests(ITestOutputHelper testOutputHelper) : MvcTestFixture
```
Убери суффикс `_Test` из названий тестов, он не добавляет информации, но удлиняет и без того немаленькое название

NikolayKorsakov replied:
+

litichevskiydv commented on README.md:
````diff
@@ -15,17 +15,35 @@ dotnet add package Byndyusoft.ApiClient
+
+2. Create your API client class and derive it from BaseClient class, inmplement your contract, if you want some specific formatter, we recomend to add it here:
+```csharp
+public class SomeApiClient : BaseClient, ISomethingGetter
````
Название клиента и название контракта должны соответствовать друг другу, скорректируй `ISomethingGetter`

NikolayKorsakov replied:
+

litichevskiydv replied:
Изначально название API клиента было верным, `SomeApiClient`. Нужно скорректировать название контракта `ISomethingGetter`, чтобы они друг другу соответствовали. Например, `ISomeApi`.

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -1,44 +1,52 @@
+        (
+            HttpClient client,
+            IOptions<ApiClientSettings> apiSettings,
+            IOptions<MediaTypeFormatter>? formatter = null
```
Зачем здесь `IOptions<>` ?

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -1,44 +1,52 @@
 
-        protected BaseClient(HttpClient client, IOptions<ApiClientSettings> apiSettings)
+        protected BaseClient
+        (
```
Перенести `(` на предыдущую строку

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -1,44 +1,52 @@
             Client = client ?? throw new ArgumentNullException(nameof(client));
             ApiSettings = apiSettings.Value ?? throw new ArgumentNullException(nameof(apiSettings));
+            Formatter = formatter?.Value ?? new JsonMediaTypeFormatter(JsonDefaults.SerializerOptions);
+            foreach (var mediaTypeHeaderValue in Formatter.SupportedMediaTypes)
```
Добавить отступ перед строкой, чтобы отделить друг от друга разные кодовые блоки

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -1,44 +1,52 @@
+            foreach (var mediaTypeHeaderValue in Formatter.SupportedMediaTypes)
+                Client.DefaultRequestHeaders.Accept.Add
+                (
+                    new MediaTypeWithQualityHeaderValue
```
Записать создание объекта в одну строку

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -1,44 +1,52 @@
+            Formatter = formatter?.Value ?? new JsonMediaTypeFormatter(JsonDefaults.SerializerOptions);
+            foreach (var mediaTypeHeaderValue in Formatter.SupportedMediaTypes)
+                Client.DefaultRequestHeaders.Accept.Add
+                (
```
Перенести `(` на предыдущую строку

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -50,64 +58,71 @@ protected Task<TResult> PostAsync<TResult>(string url, object content, Cancellat
+                .Content
+                .ReadAsAsync<TResult>
+                (
+                    new[]
```
Записать создание массива в одну строку

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/BaseClient.cs:
```diff
@@ -50,64 +58,71 @@ protected Task<TResult> PostAsync<TResult>(string url, object content, Cancellat
+            var result = await response
+                .Content
+                .ReadAsAsync<TResult>
+                (
```
Перенести `(` на предыдущую строку

NikolayKorsakov replied:
+

litichevskiydv commented on src/Byndyusoft.ApiClient/HttpContentExtensions.cs:
```diff
@@ -2,31 +2,19 @@
     using System.Threading.Tasks;
-    using Newtonsoft.Json;
 
     public static class HttpContentExtensions
```
Эти расширения все еще нужны?

NikolayKorsakov replied:
Могут быть использоаны где-то, где установлен пакет.

litichevskiydv replied:
Давай все-таки дропнем и апнем мажорик пакета

NikolayKorsakov replied:
+

NikolayKorsakov committed:
PR comments based refactoring

NikolayKorsakov committed:
client adding fixed

litichevskiydv commented on README.md:
```diff
@@ -15,17 +15,35 @@ dotnet add package Byndyusoft.ApiClient
-	public SomeApiClient(HttpClient client, IOptions<ApiClientSettings> apiSettings) : base(client, apiSettings)
-    	{
+	public SomethingGetter(HttpClient client, IOptions<ApiClientSettings> apiSettings) :
+		base(client, apiSettings, Options.Create(new SomeMediaTypeFormatter()))
```
После изменения сигнатуры конструкция `Options.Create(` не нужна

litichevskiydv commented on example/Api.Client/PersonApi.cs:
```diff
@@ -0,0 +1,45 @@
+    public PersonApi(
+            HttpClient client,
+            IOptions<ApiClientSettings> apiSettings,
+            MediaTypeFormatter? formatter = null
```
Как сейчас настроен сервер? Можно любой форматер из имеющихся передавать?

NikolayKorsakov replied:
Да, сервер принимает и отдает json, protoBuff и messagePack.

litichevskiydv commented on example/Api.Client/PersonApi.cs:
```diff
@@ -0,0 +1,45 @@
+            IOptions<ApiClientSettings> apiSettings,
+            MediaTypeFormatter? formatter = null
+        )
+        : base(client, apiSettings, formatter) { }
```
Перенести на предыдущую строчку

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api.Client/Registration/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,34 @@
+                .Configure<ApiClientSettings>(settingsSection);
+
+            services.AddHttpClient<IPersonApi, PersonApi>(
+                client=>
```
Отформатировать фрагмент

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,136 @@
+﻿namespace Byndyusoft.ApiClient.Example.Server.Controllers;
```
Отформатироовать файл

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api/Infrastructure/OpenTelemetryExtensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,46 @@
+        string? serviceName,
+        Action<OtlpExporterOptions> configureOtlp,
+        Action<TracerProviderBuilder>? configureBuilder = null,
+        Action<MeterProviderBuilder>? configureMeter = null)
```
Перенести `)` на следующую строку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api/Infrastructure/Swagger/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+        ///     AddSwagger
+        /// </summary>
+        public static IServiceCollection AddSwagger(this IServiceCollection services) =>
+                services
```
Поставить `=>` перед `services`

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api/Infrastructure/Versioning/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        public static IServiceCollection AddVersioning(this IServiceCollection services)
+        {
+            return services.AddApiVersioning(options =>
```
Перенести `options =>` на следующую строку

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api/Infrastructure/Versioning/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+                    options.ReportApiVersions = true;
+                    options.DefaultApiVersion = ApiVersion.Default;
+                    options.AssumeDefaultVersionWhenUnspecified = true;
+                })
```
Перенести `)` на следующую строку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api/Infrastructure/Versioning/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+                    options.AssumeDefaultVersionWhenUnspecified = true;
+                })
+                .AddMvc()
+                .AddApiExplorer(options =>
```
Перенести `options =>` на следующую строку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on example/Api/Infrastructure/Versioning/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,25 @@
+                {
+                    options.GroupNameFormat = "'v'VVV";
+                    options.SubstituteApiVersionInUrl = true;
+                })
```
Перенести `)` на следующую строку и отформатировать

NikolayKorsakov replied:
+

litichevskiydv commented on example/Tests/JsonFormatterTests.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public JsonFormatterTests(ITestOutputHelper testOutputHelper) : base(testOutputHelper)
+    {
+        TestSubject = new PersonApi(
```
Записать в одну строку

NikolayKorsakov replied:
+

NikolayKorsakov committed:
Projects renaming

NikolayKorsakov committed:
disposed of Startup.cs
benchmark stopped working
possible solutions
https://github.com/aspnet/Benchmarks/tree/main
https://github.com/dotnet/crank

NikolayKorsakov committed:
Отбил отступами разные смысловые блоки

NikolayKorsakov committed:
readme update

NikolayKorsakov committed:
pr comments based formatting

NikolayKorsakov committed:
benchmark fix

litichevskiydv commented on example/Api.Client/PersonApi.cs:
```diff
@@ -0,0 +1,44 @@
+
+public class PersonApi : BaseClient, IPersonApi
+{
+    public PersonApi(
```
Переименовать `PersonApi` => `PersonApiClient`

myasnik0v replied:
+

litichevskiydv commented on example/Api.Client/Api.Client.csproj:
```diff
@@ -0,0 +1,20 @@
+        <TargetFramework>net8.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+        <RootNamespace>Api.Client</RootNamespace>
```
Через пропсы настроить `RootNamespace`

litichevskiydv commented on example/Api.Contracts/Api.Contracts.csproj:
```diff
@@ -0,0 +1,14 @@
+        <TargetFramework>net8.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+        <RootNamespace>Api.Contracts</RootNamespace>
```
Через пропсы настроить `RootNamespace`

litichevskiydv commented on example/Api/Api.csproj:
```diff
@@ -0,0 +1,36 @@
+	<PropertyGroup>
+		<OutputType>Exe</OutputType>
+		<TargetFramework>net8.0</TargetFramework>
+		<RootNamespace>Api</RootNamespace>
```
Через пропсы настроить `RootNamespace`

litichevskiydv commented on example/Api/Controllers/PersonController.cs:
```diff
@@ -0,0 +1,136 @@
+
+[ApiController]
+[Route("[controller]")]
+public class PersonController : ControllerBase
```
Переимновать `PersonController` => `PersonApiController `

myasnik0v replied:
А зачем суффикс API? Мб Person**s**Controller

litichevskiydv commented on example/Api/Program.cs:
```diff
@@ -0,0 +1,66 @@
+                .WriteToOpenTelemetry(activityEventBuilder: StructuredActivityEventBuilder.Instance)
+    );
+
+var services = builder.Services;
```
Использоватьь FluentSyntax там, где можно, и подбить отступы

litichevskiydv commented on example/Benchmark/Benchmark.csproj:
```diff
@@ -0,0 +1,25 @@
+    <PropertyGroup>
+        <TargetFramework>net8.0</TargetFramework>
+        <OutputType>Exe</OutputType>
+        <RootNamespace>Benchmark</RootNamespace>
```
При помощи пропсов настроить `RootNamespace`

litichevskiydv commented on example/Benchmark/MvcTestFixture.cs:
```diff
@@ -0,0 +1,45 @@
+            .WithWebHostBuilder(
+                builder =>
+                {
+                    builder.UseTestServer();
```
Проверить, нужно ли это. Как будто, без этой строчки можно обойтись

litichevskiydv commented on example/Benchmark/MvcTestFixture.cs:
```diff
@@ -0,0 +1,45 @@
+                }
+            )
+            .CreateClient(
+                new WebApplicationFactoryClientOptions
```
Записать в одну строку

litichevskiydv commented on example/Models/Models.csproj:
```diff
@@ -0,0 +1,14 @@
+        <TargetFramework>net8.0</TargetFramework>
+        <ImplicitUsings>enable</ImplicitUsings>
+        <Nullable>enable</Nullable>
+        <RootNamespace>Models</RootNamespace>
```
При помощи пропсов настроить RootNamespace

litichevskiydv commented on example/Models/PersonModel.cs:
```diff
@@ -0,0 +1,98 @@
+        bool? isMarried = null,
+        FavoriteDessertEnum? favoriteDessert = null,
+        ulong id = 0,
+        ulong? driverLicenseId = null)
```
Перенести `)` на следующую строку и отформатировать

litichevskiydv commented on example/Models/PersonModel.cs:
```diff
@@ -0,0 +1,98 @@
+        111
+    );
+
+    public override int GetHashCode() =>
```
Перенести `=>` на следующую строку и отформатировать

litichevskiydv commented on example/Tests/Infrastructure/CustomWebApplicationFactory.cs:
```diff
@@ -0,0 +1,15 @@
+using Microsoft.AspNetCore.Mvc.Testing;
+using Microsoft.AspNetCore.TestHost;
+
+public class CustomWebApplicationFactory<TProgram>
```
Разобраться, нужен ли этот класс. Возможно, от него можно отказаться

litichevskiydv commented on example/Tests/MvcTestFixture.cs:
```diff
@@ -0,0 +1,34 @@
+    protected IOptions<ApiClientSettings> _clientSettings = new OptionsWrapper<ApiClientSettings>(
+        new ApiClientSettings
+        {
+            ConnectionString = "http://localhost:5000"
```
Эту настройку нужно брать из `HttpClient`, который возвразает `WebApplicationFactory`

litichevskiydv commented on example/Tests/Tests.csproj:
```diff
@@ -0,0 +1,25 @@
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <IsPackable>false</IsPackable>
+    <RootNamespace>Tests</RootNamespace>
```
Настроить `RootNamespace` при помощи пропсов
